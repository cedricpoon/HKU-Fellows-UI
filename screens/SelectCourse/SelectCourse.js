import React, { Component } from 'react';
import { Container, Content } from 'native-base';

import { Header, CourseScrollable } from 'hkufui/components'
import { localize } from 'hkufui/locale';
import courseList from 'hkufui/mock/private/courses.json';

const locale = localize({ language: 'en', country: 'hk' });

class SelectCourse extends Component {

  render() {
    return (
      <Container>
        <Header title={locale['header.course']} backable />
        <Content padder>
          <CourseScrollable list={courseList} />
        </Content>
      </Container>
    );
  }
}

export default SelectCourse;
