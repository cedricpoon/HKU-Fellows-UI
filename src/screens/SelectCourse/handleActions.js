import * as types from "hkufui/src/constants/actionTypes";

export const onUpdateLocation = ({item}) => ({
  type: types.UPDATE_LOCATION,
  courseId: item.id,
  courseTitle: item.title
});

export const onSetSelectCourseIndex = (array) => ({
  type: types.SET_SELECT_COURSE_INDEX,
  breadcrumb: array
});

const handleActions = (state = {}, action = {}) => {
  switch (action.type) {
    case types.UPDATE_LOCATION:
      return {
        ...state,
        courseId: action.courseId,
        courseTitle: action.courseTitle
      }
    case types.SET_SELECT_COURSE_INDEX:
      return {
        ...state,
        breadcrumb: action.breadcrumb
      }
    default:
      return state;
  }
};

export default handleActions;
