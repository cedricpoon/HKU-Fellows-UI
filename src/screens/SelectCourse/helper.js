export const formBreadcrumbString = (breadcrumbArray) => {
  if (breadcrumbArray.length === 0)
    return;

  var retn = '';
  const seperator = ' / ';

  breadcrumbArray.forEach((item) => {
    if (retn !== '')
      retn += seperator;
    retn += item.title
  });

  return retn;
}
