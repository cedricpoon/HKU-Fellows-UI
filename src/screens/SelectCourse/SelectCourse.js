import React, { Component } from 'react';
import { Container, Content } from 'native-base';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Header, CourseScrollable } from 'hkufui/components'
import { localize } from 'hkufui/locale';
import courses from 'hkufui/mock/public/courses';

import { onUpdateLocation, onSetSelectCourseIndex } from './handleActions';
import { fetchPostsSafe } from '../Preview/PostPreviewLoader/loadPosts';
import { formBreadcrumbString } from './helper';

const locale = localize({ language: 'en', country: 'hk' });

export class SelectCourse extends Component {
  componentDidUpdate() {
    const { breadcrumb } = this.props;

    // check if header did mount
    if (this._header)
      this._header.setSubtitle(
        formBreadcrumbString(breadcrumb)
      );
  }

  componentDidMount() {
    this.componentDidUpdate();
  }

  render() {
    const {
      courses,
      onUpdateLocationAndLoadPost,
      onSetSelectCourseIndex,
      breadcrumb
    } = this.props;

    const onItemPressWrapper = ({item}) => {
      return () => {
        onUpdateLocationAndLoadPost(item);
      };
    };

    return (
      <Container>
        <Header
          title={locale['header.course']}
          backable
          rightIcon='arrow-dropup'
          onRightPress={() => {
            this._courseScrollable.collapseAll();
          }}
          ref={ref => this._header = ref}
        />
        <Content padder>
          <CourseScrollable
            list={courses}
            onItemPressWrapper={onItemPressWrapper}
            onSetSelectCourseIndex={onSetSelectCourseIndex}
            ref={ref => this._courseScrollable = ref}
            expandedList={breadcrumb}
          />
        </Content>
      </Container>
    );
  }
}

/* Mocking mapStateToProps */
SelectCourse.defaultProps = {
  courses: courses,
  breadcrumb: []
}

SelectCourse.propTypes = {
  courses: PropTypes.arrayOf(PropTypes.object).isRequired,
  onUpdateLocationAndLoadPost: PropTypes.func.isRequired,
  onSetSelectCourseIndex: PropTypes.func.isRequired,
  breadcrumb: PropTypes.array
}

const mapDispatchToProps = dispatch => ({
  onUpdateLocationAndLoadPost: (item) => dispatch(fetchPostsSafe(onUpdateLocation(item))),
  onSetSelectCourseIndex: (array) => dispatch(onSetSelectCourseIndex(array))
})

const mapStateToProps = state => ({
  breadcrumb: state.location.breadcrumb
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SelectCourse);
