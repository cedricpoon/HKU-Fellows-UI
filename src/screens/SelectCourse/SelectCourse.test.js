import React from 'react';
import { shallow } from 'enzyme';
import { SelectCourse } from './SelectCourse';

describe('Testing SelectCourse Screen', () => {
  const defaultProps = {
    courses: [],
    onUpdateLocationAndLoadPost: () => {},
    onSetSelectCourseIndex: () => {}
  }

  it('renders as expected with defaultProps', () => {
    const wrapper = shallow(
      <SelectCourse {...defaultProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as expected with defaultProps and breadcrumb', () => {
    const wrapper = shallow(
      <SelectCourse {...defaultProps} breadcrumb={[]} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
