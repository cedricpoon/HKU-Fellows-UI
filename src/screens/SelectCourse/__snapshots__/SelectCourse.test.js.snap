// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SelectCourse Screen renders as expected with defaultProps 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectCourse
    breadcrumb={Array []}
    courses={Array []}
    onSetSelectCourseIndex={[Function]}
    onUpdateLocationAndLoadPost={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <withNavigation(Header)
          backable={true}
          onRef={[Function]}
          onRightPress={[Function]}
          rightIcon="arrow-dropup"
          title="Courses"
        />,
        <Styled(Content)
          padder={true}
        >
          <CourseScrollable
            expandedList={Array []}
            list={Array []}
            onItemPressWrapper={[Function]}
            onSetSelectCourseIndex={[Function]}
          />
        </Styled(Content)>,
      ],
      "virtual": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "backable": true,
          "onRef": [Function],
          "onRightPress": [Function],
          "rightIcon": "arrow-dropup",
          "title": "Courses",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <CourseScrollable
            expandedList={Array []}
            list={Array []}
            onItemPressWrapper={[Function]}
            onSetSelectCourseIndex={[Function]}
          />,
          "padder": true,
          "virtual": undefined,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "expandedList": Array [],
            "list": Array [],
            "onItemPressWrapper": [Function],
            "onSetSelectCourseIndex": [Function],
          },
          "ref": [Function],
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <withNavigation(Header)
            backable={true}
            onRef={[Function]}
            onRightPress={[Function]}
            rightIcon="arrow-dropup"
            title="Courses"
          />,
          <Styled(Content)
            padder={true}
          >
            <CourseScrollable
              expandedList={Array []}
              list={Array []}
              onItemPressWrapper={[Function]}
              onSetSelectCourseIndex={[Function]}
            />
          </Styled(Content)>,
        ],
        "virtual": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "backable": true,
            "onRef": [Function],
            "onRightPress": [Function],
            "rightIcon": "arrow-dropup",
            "title": "Courses",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <CourseScrollable
              expandedList={Array []}
              list={Array []}
              onItemPressWrapper={[Function]}
              onSetSelectCourseIndex={[Function]}
            />,
            "padder": true,
            "virtual": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "expandedList": Array [],
              "list": Array [],
              "onItemPressWrapper": [Function],
              "onSetSelectCourseIndex": [Function],
            },
            "ref": [Function],
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Testing SelectCourse Screen renders as expected with defaultProps and breadcrumb 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectCourse
    breadcrumb={Array []}
    courses={Array []}
    onSetSelectCourseIndex={[Function]}
    onUpdateLocationAndLoadPost={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <withNavigation(Header)
          backable={true}
          onRef={[Function]}
          onRightPress={[Function]}
          rightIcon="arrow-dropup"
          title="Courses"
        />,
        <Styled(Content)
          padder={true}
        >
          <CourseScrollable
            expandedList={Array []}
            list={Array []}
            onItemPressWrapper={[Function]}
            onSetSelectCourseIndex={[Function]}
          />
        </Styled(Content)>,
      ],
      "virtual": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "backable": true,
          "onRef": [Function],
          "onRightPress": [Function],
          "rightIcon": "arrow-dropup",
          "title": "Courses",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <CourseScrollable
            expandedList={Array []}
            list={Array []}
            onItemPressWrapper={[Function]}
            onSetSelectCourseIndex={[Function]}
          />,
          "padder": true,
          "virtual": undefined,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "expandedList": Array [],
            "list": Array [],
            "onItemPressWrapper": [Function],
            "onSetSelectCourseIndex": [Function],
          },
          "ref": [Function],
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <withNavigation(Header)
            backable={true}
            onRef={[Function]}
            onRightPress={[Function]}
            rightIcon="arrow-dropup"
            title="Courses"
          />,
          <Styled(Content)
            padder={true}
          >
            <CourseScrollable
              expandedList={Array []}
              list={Array []}
              onItemPressWrapper={[Function]}
              onSetSelectCourseIndex={[Function]}
            />
          </Styled(Content)>,
        ],
        "virtual": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "backable": true,
            "onRef": [Function],
            "onRightPress": [Function],
            "rightIcon": "arrow-dropup",
            "title": "Courses",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <CourseScrollable
              expandedList={Array []}
              list={Array []}
              onItemPressWrapper={[Function]}
              onSetSelectCourseIndex={[Function]}
            />,
            "padder": true,
            "virtual": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "expandedList": Array [],
              "list": Array [],
              "onItemPressWrapper": [Function],
              "onSetSelectCourseIndex": [Function],
            },
            "ref": [Function],
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
